Selectors reworked ... (#98)
ietf-wg-jsonpath/draft-ietf-jsonpath-base: draft-ietf-jsonpath-base.md

## Selectors

A JSONPath query consists of a sequence of selectors. Valid selectors are

  * Root selector `'$'`
  * Dot selector `'.<name>'`, used with object member names exclusively.
  * Dot wild card selector `'.*'`.
  * Index selector `'[<index>]'`, where `<index>` is either an (possibly negative) array index or an object member name.
  * Index wild card selector `'[*]'`.
  * Array slice selector `'[<start>:<end>:<step>]'`, where `<start>`, `<end>`, `<step>` are integer literals.
  * Nested descendants selector `'..'`.
  * Union selector `'[<sel1>,<sel2>,...,<selN>]'`, holding a comma delimited list of selectors.
  * Filter selector `[?(<expr>)]`

>   * Current item selector `'@'`

Reviewer: Resolved in #102

### Root Selector

#### Syntax
{: numbered="false" toc="exclude"}

Every valid JSONPath query MUST begin with the root selector named `'$'`.

~~~~ abnf
root-selector  = %x24                 ; $
~~~~

#### Semantics
The Argument &mdash; the root JSON value &ndash; is anonymous by nature. By getting assigned the universal name `'$'` it becomes the root node. 

### Dot Selector

#### Syntax
{: numbered="false" toc="exclude"}

A dot selector starts with a dot `'.'` followed by an object's member name.

~~~~ abnf
dot-selector    = "." dot-member-name
dot-member-name = 1*(
                      DIGIT /
                      ALPHA /
                      "_"   /           ; _
                      %x80-10FFFF       ; any non-ASCII Unicode character
                    )
DIGIT           =  %x30-39              ; 0-9

> ALPHA           =  %x41-5A / %x61-7A    ; A-Z / a-z

Author: I still needs to be discussed, if a dot-member-name is allowed to start with a DIGIT.

> ALPHA           =  %x41-5A / %x61-7A    ; A-Z / a-z

Reviewer: What does applying `.1` to `[1, 2, 3]` mean?

~~~~

> Member names containing other ASCII characters than allowed by `dot-selector` &ndash; especially space `' '` and minus `'-'` characters MUST NOT be used with the `dot-selector`. Member names containing these might be addressed by the `index-selector` instead. 

Author: Yes, as an autor hint it should be moved elsewhere.

> Member names containing other ASCII characters than allowed by `dot-selector` &ndash; especially space `' '` and minus `'-'` characters MUST NOT be used with the `dot-selector`. Member names containing these might be addressed by the `index-selector` instead. 

Reviewer: (This is not just about other ASCII characters, but other characters in general.)

> Member names containing other ASCII characters than allowed by `dot-selector` &ndash; especially space `' '` and minus `'-'` characters MUST NOT be used with the `dot-selector`. Member names containing these might be addressed by the `index-selector` instead. 

Reviewer: MAY is an interoperability keyword (RFC 2119), while this is just a suggestion for an alternative approach, so MAY seems inappropriate.  Proposed wording in #102.

#### Semantics
{: numbered="false" toc="exclude"}

The `dot-selector` selects the value corresponding to the member name from any JSON object. It selects no values from any other JSON value.

Note that the `dot-selector` follows the philosophy of JSON strings and is
allowed to contain bit sequences that cannot encode Unicode characters (a
single unpaired UTF-16 surrogate, for example).
The behaviour of an implementation is undefined for member names which do

> not encode Unicode characters.

Reviewer: Cannot happen.

### Dot Wild Card Selector

#### Syntax
{: numbered="false" toc="exclude"}

> The dot wild card selector has the shape `'.*'`.

Reviewer: Resolved in #102

~~~~ abnf
dot-wild-selector    = "." "*"            ;  dot followed by asterisk
~~~~

#### Semantics
{: numbered="false" toc="exclude"}

> A dot selector combined with a single asterisk is a wild card. It selects all member values of an object as well as all elements of an array. Applying the `dot-wild-selector` to primitive JSON values, as number, string, or literal, does nothing.

Reviewer: Resolved in #102

### Index Selector

#### Syntax
{: numbered="false" toc="exclude"}

An index selector `'[<index>]'` addresses a single object member value or a single array element value.

~~~~ abnf
index-selector      = "[" (quoted-member-name / element-index) "]"
~~~~

Applying the `index-selector` to an object value, a `quoted-member-name` string is required. JSONPath allows it to be enclosed in _single_ or _double_ quotes.

~~~~ abnf
quoted-member-name  =  %x22 *double-quoted %x22 /       ; "string"

>                        %x27 *single-quoted %x27         ; 'string'

Author: ' simply means multiple ...

>                        %x27 *single-quoted %x27         ; 'string'

Reviewer: This should point to the string literal syntax that we are using for comparisons as well.

escapable           = (
                          b /         ;  BS backspace U+0008
                          t /         ;  HT horizontal tab U+0009
                          n /         ;  LF line feed U+000A
                          f /         ;  FF form feed U+000C
                          r /         ;  CR carriage return U+000D
                          / /         ;  /  slash (solidus)
                          \ /         ;  \  backslash (reverse solidus)
                          u 4HEXDIG   ;  uXXXX      U+XXXX
                      )
HEXDIG              = %x41-46 /           ;  A-F
                      %x61-66             ;  a-f
~~~~

Applying the `index-selector` to an array, a numerical `element-index` is required. JSONPath allows it to be negative.

~~~~ abnf
element-index   = int                             ; decimal integer

int             = ["-"] ( "0" / (DIGIT1 *DIGIT) ) ; -  optional 
DIGIT1          = %x31-39                         ; 1-9 non-zero digit
~~~~

Notes:
1. `double-quoted` strings follow JSON in {{RFC8259}}.
   `single-quoted` strings follow an analogous pattern.
2. An `element-index` is an integer (in base 10).

> 3. The syntax does not allow octal-like integers with leading zeros such as `01` or `-01`.

Reviewer: That is again something that should be done in the expression language.

| \\"                | U+0022              | quotation mark              |
| \\'                | U+0027              | apostrophe                  |
| \\/                | U+002F              | slash (solidus)             |
| \\\\               | U+005C              | backslash (reverse solidus) |
| \\uXXXX            | U+XXXX              | unicode character           |

{: title="Escape Sequence Replacements" cols="c c"}

The `index-selector` applied with a `quoted-member-name` to an object selects the corresponding value from it, if and only if that object has a member with that name. 
Nothing is selected from a value which is not a object.

Array indexing via `element-index` is a way of selecting a particular array element using a zero-based index.
For example, selector `[0]` selects the first and selector `[4]` the fifth element of a sufficiently filled array.

A negative `element-index` counts from the array end.
For example, selector `[-1]` selects the last and selector `[-2]` selects the last but one element of an array with at least two elements.

### Index Wild Card Selector

#### Syntax
{: numbered="false" toc="exclude"}

> The index wild card selector has the shape `'[*]'`.

Reviewer: Resolved in #102

~~~~ abnf
index-wild-selector    = "[" "*" "]"  ;  asterisk enclosed by brackets
~~~~

#### Semantics
{: numbered="false" toc="exclude"}

> An index selector combined with a single asterisk is a wild card. It selects all member values of an object as well as all elements of an array. Applying the `index-wild-selector` to primitive JSON values, as number, string, or literal, does nothing. The `index-wild-selector` behaves identically to the `dot-wild-selector`.

Reviewer: Resolved in #102

### Array Slice Selector

#### Syntax
{: numbered="false" toc="exclude"}

> The array slice selector has the shape `'[<start>:<end>:<step>]'`. It selects elements starting at index `<start>`, ending at &ndash; not including &ndash; `<end>`, while incrementing by `step`.

Reviewer: Resolved in #102

~~~~ abnf
slice-selector = "[" slice-index "]"

> slice-index    = ws [start] ws ":" ws [end] [ws ":" ws [step] ws]

Author: I would also like to drop them. Is there a general recommended way to handle them in ABNF ?

start          = int       ; included in selection
end            = int       ; not included in selection
step           = int       ; default: 1 

ws             = *( %x20 / ; Space
                    %x09 / ; Horizontal tab
                    %x0A / ; Line feed or New line
                    %x0D ) ; Carriage return
~~~~

> The `slice-selector` consists of three optional decimal integers separated by colons.

Author: I believe the second colon is already optional ...

> The `slice-selector` consists of three optional decimal integers separated by colons.

Reviewer 2: Yes, I see in the syntax graph.  Maybe this sentence could be updated, or perhaps even keep the note that this line replaces.  I think having these examples is useful.

fit in the implementation's representation of an integer, the implementation
MUST raise an error.

### Descendant Selector

#### Syntax
{: numbered="false" toc="exclude"}

> The descendant selector starts with a double dot `'..'`. It can be used in place of the normal `dot-selector`. So it will usually be followed by an object member name. As an extension to that natural rule it might also be followed by an `index-selector` acting on objects or arrays.

Author: ... this is nearly the wording in E4X - where I stole that operator from. This is quite important:

* Combination of descendant-operator with index operator is valid: `..['key']` as is `..[*]`
* But combination of descendant-operator with dot-operator resulting in `...key` or `...*` is not.

So formulation "can be used in place of the normal `dot-selector`" makes this clear, in order to avoid the latter.

> The descendant selector starts with a double dot `'..'`. It can be used in place of the normal `dot-selector`. So it will usually be followed by an object member name. As an extension to that natural rule it might also be followed by an `index-selector` acting on objects or arrays.

Reviewer 2: The `descendant-selector` and `dot-selector` are completely different selectors that happen to have an overlapping syntax.  After all, `$..['foo']` is valid, but `$.['foo']` is not.

To say that `descendant-selector` can be used "in place of" `dot-selector` conveys replacement, which is incorrect because they do different things.  It's more correct to say that they share a similar syntax in that they both support `dot-member-name`.

To be honest, I'm not sure we need the sentence at all.

```suggestion
The descendant selector starts with a double dot `'..'` and can be followed by an object member name (similar to the `dot-selector`) or by an `index-selector` acting on objects or arrays.
```

> The descendant selector starts with a double dot `'..'`. It can be used in place of the normal `dot-selector`. So it will usually be followed by an object member name. As an extension to that natural rule it might also be followed by an `index-selector` acting on objects or arrays.

Reviewer: Resolved in #102

(Of course, we can, and probably will, tweak further.)

~~~~ abnf
descendant-selector = ".." ( dot-member-name      /  ; ..<name>
                             index-selector       /  ; ..[<index>]
                             index-wild-selector  /  ; ..[*]
                             "*"                     ; ..*
                           )
~~~~

#### Semantics
{: numbered="false" toc="exclude"}

The `descendant-selector` is inspired by ECMAScript for XML (E4X). It selects object members and array elements in a recursive descent process.

### Union Selector

#### Syntax

The union selector is syntactically related to the `index-selector`. It contains multiple, comma separated entries.

~~~~ abnf
union-selector = "[" ws union-entry 1*(ws "," ws union-enty) ws "]"

union-entry    =  ( quoted-member-name / 
                    element-index      /
                    slice-index
                  )
~~~~

#### Semantics

A union selects any node which is selected by at least one of the union selectors and selects the concatenation of the
lists (in the order of the selectors) of nodes selected by the union elements.

> Possibly duplicate selected nodes are kept in the node list.

Reviewer 3: ```suggestion
Any duplicated selected nodes are kept in the node list.
```

Reviewer: Define word duplicate (or preferable don't use it but describe what happens).

Reviewer: Resolved in #102

### Filter Selector

#### Syntax

> The filter selector has the shape `'[?<expr]'`. It works via iterating over container values, i.e. arrays and objects. 

Author: ... good catch ..

~~~~ abnf
filter-selector    = "[?" boolean-expr "]"
~~~~

During iteration process each array element or object member is visited and its value &ndash; accessible via symbol `'@'` &ndash; or one of its descendants &ndash; uniquely defined by a relative path &ndash; is tested against a boolean expression `boolean-expr`.

> If the result is `true`, the current item, represented by `'@'`, is selected. In case of a `false` result, it is not.

Reviewer 3: ```suggestion
The current item is selected if and only if the result is `true`.
```

Reviewer 2: This change properly captures whether the item should be selected if the expression fails evaluation (e.g. if `@.foo` isn't found).

Reviewer: Resolved in #102

~~~~ abnf
boolean-expr = *logical-expr
logical-expr = [neg-op] ["("] comp-expr *[logical-op comp-expr] [")"]
neg-op       = "!"                                  ; not operator
logical-op   = "||" / "&&"                          ; logical operator
comp-expr    = (rel-path-val /
                calc-val) [(comp-op comparable /  ; comparison
                            regex-op regex     /  ; RegEx test

>                             in-op container )]    ; containment test

Author: yes ...

>                             in-op container )]    ; containment test

Reviewer: `array` and `object` don't have a definition either.
Do you mean expressions that return an array or object?
Or maybe just literals of these types?

>                             in-op container )]    ; containment test

Author: both ...

`[?@.color in ['red,'green','blue']`

or 

`[?@.color in $.validColors]`

comp-op      = "==" / "!=" /                        ; comparison ...
               "<"  / ">"  /                        ; operators
               "<=" / ">="
regex-op     = "~="                                 ; RegEx match
in-op        = " in "                               ; in operator
comparable   = number / quoted-string /             ; primitive ...
               true / false / null /                ; values only
               rel-path-val /                       ; descendant value
               calc_val /                           ; calculated value
               json-path                            ; any value
               
rel-path-val = "@" *(dot-selector / index-selector) 

> calc_val     = func "(" [rel-path-val / json-path] ")"

Author: in fact i am quite reluctant here. It's a proposal needing discussion ... will most certainly result in an issue..

func         = "index"
~~~~

Notes:

> * Parenthesis can be used with `boolean-expr` for grouping. So filter selection syntax in the original proposal `'[?(<expr)]'` is naturally contained in the current lean syntax `'[?<expr]'` as a special case.

Author: this is still a proposal ... I see no possible syntax collisons ... it is backward compatibel and lean.

> * Parenthesis can be used with `boolean-expr` for grouping. So filter selection syntax in the original proposal `'[?(<expr)]'` is naturally contained in the current lean syntax `'[?<expr]'` as a special case.

Reviewer 2: I'd like to see a set of examples.  The parentheses create a level of human-readability that I'm afraid will be lost by removing them.

> * Parenthesis can be used with `boolean-expr` for grouping. So filter selection syntax in the original proposal `'[?(<expr)]'` is naturally contained in the current lean syntax `'[?<expr]'` as a special case.

Reviewer: Note that the parentheses can still be a convention, even if we don't require them.

> * Parenthesis can be used with `boolean-expr` for grouping. So filter selection syntax in the original proposal `'[?(<expr)]'` is naturally contained in the current lean syntax `'[?<expr]'` as a special case.

Author: yes ... I've seen this notation with some implementations in the wild and .. as a minimalist .. I liike it.

* Comparisons are restricted to primitive values `number`, `string`, `true`, `false`, `null`. Comparisons with complex values will fail, i.e. no selection occurs.

> * Implicite type conversions during comparisons are not performed. So `"13 == '13'"` selects nothing.

Reviewer: We probably should not bury the details of the expression language in the selector description.

We cannot notate complex (structured, non-atom, ...) values so I don't know how that clause would become active.

s/Parenthesis/Parentheses/

> * Implicite type conversions during comparisons are not performed. So `"13 == '13'"` selects nothing.

Reviewer 2: I would say that we can notate them just fine.  An expression should be able to operate on any JSON literal.  I see no reason why `@.foo == [1, 2]` should be disallowed.

> * Implicite type conversions during comparisons are not performed. So `"13 == '13'"` selects nothing.

Reviewer: Resolved in #102

> * Implicite type conversions during comparisons are not performed. So `"13 == '13'"` selects nothing.

Author: We can restrict comparisons to simple values only (some say 'scalars`) . You can write here
```
[? @.foo[0] == 1 && @.foo.[1] == 2]
```
instead.

Benefit in use seems to be low, but complexity might become huge.

> * A member or element value by itself is *falsy* only, if it does not exist. Otherwise it is *truthy*, resulting in its value. To be more specific explicite comparisons are necessary. This existence test &ndash; as an exception of the general rule &ndash; also works with complex values.

Reviewer 3: Please define the terms _truthy_ and _falsy_.

Author: Instead define it, dropping is better.

> * A member or element value by itself is *falsy* only, if it does not exist. Otherwise it is *truthy*, resulting in its value. To be more specific explicite comparisons are necessary. This existence test &ndash; as an exception of the general rule &ndash; also works with complex values.

Reviewer: Resolved in #102

* Regular expression tests can be applied to `string` values only.
* Containment tests work with arrays and objects.

> * Explicite boolean type conversion is done by the not operator `neg-op`.

Reviewer 3: ```suggestion
* Explicit boolean type conversion is done by the not operator `neg-op`.
```

Reviewer: Resolved in #102

> * Explicite boolean type conversion is done by the not operator `neg-op`.

Reviewer 3: ```suggestion
* Explicit boolean type conversion is done by the not operator `neg-op`.
```
but what does this mean?

Author: ... look at the table near by. Converting numbers, strings to booleans, as done in C language.

Reviewer: Resolved in #102

> * The behaviour of operators is consistent with the 'C'-family of programming languages.

Reviewer 3: This statement seems out of place in the syntax section. Do we need an "Informal Semantics" section?

Author: will be moved. Not sure when to move to semantic or informal semantics section.

Reviewer 3: @goessner  wrote:
> will be moved. Not sure when to move to semantic or informal semantics section.

 If by "when" you meant "whether", I think the statement should be moved to a new informal semantics section.

Author: It should perhaps read better:
```
Syntax and behaviour of operators is consistent with the 'C'-family of programming languages.
```
Otherwise we might need to add a table of operator precedence ... which would be no problem though.

Reviewer: FYI: Have a look at how RFC 8949 solves that problem:
https://www.rfc-editor.org/rfc/rfc8949.html#section-1.2-6

The IESG will insist on a reference to the C standard, I can tell you this much :-)

Author: @Glyn: By "when" I mean "whenever" in comparable situations, should it goto

* syntax
* semantics
* informal semantics

Are there specific rules especially to allocate to the last two.

Reviewer 3: > @glyn: By "when" I mean "whenever" in comparable situations, should it goto
> 
>     * syntax
> 
>     * semantics
> 
>     * informal semantics
> 
> 
> Are there specific rules especially to allocate to the last two.

No rules, but the idea with array slices was to introduce the semantics gently in the (informative) informal semantics section and then spell out the (normative) semantics, possibly more indigestibly, in the semantics section. But I am not wedded to this convention so long as we don't confuse normative and informative text.
